File: 22-Binary-Trees-Miscellaneous/python/05_kth_largest_element_in_an_array.py
import random
from typing import List

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        target_index = len(nums) - k
        left, right = 0, len(nums) - 1

        while left < right:
            pivot_index = random.randint(left, right)
            pivot = nums[pivot_index]

            lo, hi, i = left, right, left
            while i <= hi:
                if nums[i] < pivot:
                    nums[i], nums[lo] = nums[lo], nums[i]
                    lo += 1
                    i += 1
                elif nums[i] > pivot:
                    nums[i], nums[hi] = nums[hi], nums[i]
                    hi -= 1
                else:
                    i += 1

            if target_index < lo:
                right = lo - 1
            elif target_index > hi:
                left = hi + 1
            else:
                return pivot

        return nums[left]

File: 22-Binary-Trees-Miscellaneous/java/05-KthLargestElementInAnArray.java
import java.util.*;

class Solution {
    public int findKthLargest(int[] nums, int k) {
        int targetIndex = nums.length - k;
        return quickselect(nums, 0, nums.length - 1, targetIndex);
    }

    private int quickselect(int[] nums, int left, int right, int targetIndex) {
        while (left < right) {
            int pivotIndex = left + new Random().nextInt(right - left + 1);
            int pivot = nums[pivotIndex];

            int lo = left, hi = right, i = left;
            while (i <= hi) {
                if (nums[i] < pivot) {
                    swap(nums, i, lo);
                    lo++;
                    i++;
                } else if (nums[i] > pivot) {
                    swap(nums, i, hi);
                    hi--;
                } else {
                    i++;
                }
            }

            if (targetIndex < lo) {
                right = lo - 1;
            } else if (targetIndex > hi) {
                left = hi + 1;
            } else {
                return pivot;
            }
        }
        return nums[left];
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
